Linker babel translator

petebacondarwin opened the pull request:
refactor(compiler-cli): implement `BabelAstFactory` and `AstHost`s 

This PR adds the `AstHost` interface, along with implementations for
both Babel and TS.

It also implements the Babel vesion of the `AstFactory` interface, along
with a linker specific implementation of the `ImportGenerator` interface.

These classes will be used by the new "ng-linker" to transform prelinked
library code using a Babel plugin.

petebacondarwin committed:
refactor(compiler-cli): implement `BabelAstFactory` and `AstHost`s

This commit adds the `AstHost` interface, along with implementations for
both Babel and TS.

It also implements the Babel vesion of the `AstFactory` interface, along
with a linker specific implementation of the `ImportGenerator` interface.

These classes will be used by the new "ng-linker" to transform prelinked
library code using a Babel plugin.

JoostK commented on packages/compiler-cli/linker/src/ast/typescript/typescript_ast_host.ts:
```diff
@@ -0,0 +1,138 @@
+import {AstHost, Range} from '../ast_host';
+import {assert} from '../utils';
+
+export class TypeScriptAstHost implements AstHost<ts.Expression> {
```
So this is probably implemented to avoid bias towards Babel when considering how to define the `AstHost` implementation, but we don't currently plan on actually using this, do we?

JoostK replied:
Also, I really like how similar this is to the Babel implementation

petebacondarwin replied:
Yes and yes. I feel it is useful enough to keep it in the codebase.

JoostK commented on packages/compiler-cli/linker/src/ast/typescript/typescript_ast_host.ts:
```diff
@@ -0,0 +1,138 @@
+    for (const property of obj.properties) {
+      assert(property, ts.isPropertyAssignment, 'a property assignment');
+
+      let key: string;
+      if (ts.isIdentifier(property.name) || ts.isStringLiteral(property.name) ||
+          ts.isNumericLiteral(property.name)) {
+        key = property.name.text;
+      } else {
+        throw new FatalLinkerError(property.name, 'Unsupported syntax, expected a property name');
+      }
```
This looks like following the structure as found in the Babel implementation, but the `if`-guard could be inverted here.

petebacondarwin replied:
I'll replace with an assertion call...

petebacondarwin replied:
And do something similar in the Babel host.

JoostK commented on packages/compiler-cli/linker/test/ast/babel/babel_ast_factory_spec.ts:
```diff
@@ -0,0 +1,373 @@
+      expect(generate(literal).code).toEqual('42');
+    });
+
+    it('should create a number literal', () => {
```
Did you mean this to have an equal description to the above tests, if it's specifically targeting `NaN`?

petebacondarwin replied:
LOL - https://github.com/angular/angular/pull/38775#discussion_r491712955

JoostK commented on packages/compiler-cli/linker/test/ast/babel/babel_ast_factory_spec.ts:
```diff
@@ -0,0 +1,373 @@
+         expect(generate(varDecl).code).toEqual('const foo = 42;');
+       });
+
+    it('should create a downlevelled variable declaration statement node for the given variable name and initializer',
```
🤡 🇺🇸

petebacondarwin replied:
😭

JoostK commented on packages/compiler-cli/linker/test/ast/babel/babel_ast_host_spec.ts:
```diff
@@ -0,0 +1,282 @@
+    });
+  });
+
+  describe('parseStringLiteral()', () => {
```
Given that these tests are so complete, can we have tests for the exception case?

petebacondarwin replied:
Done.

JoostK commented on packages/compiler-cli/linker/test/ast/babel/babel_ast_host_spec.ts:
```diff
@@ -0,0 +1,282 @@
+      expect(host.isFunctionExpression(expr(template.ast`function foo() {}`))).toBe(false);
+    });
+
+    it('should return false if the expression is not an array literal', () => {
```
I see, you copy-pasted all of it! ✂️
```suggestion
    it('should return false if the expression is not a function expression', () => {
```

petebacondarwin replied:
(maybe...)

JoostK commented on packages/compiler-cli/linker/test/ast/babel/babel_ast_host_spec.ts:
```diff
@@ -0,0 +1,282 @@
+    it('should error if the body has more than 1 statement', () => {
+      expect(
+          () =>
+              host.parseReturnValue(rhs(template.ast`x = function () { const x = 10; return x;}`)))
```
```suggestion
              host.parseReturnValue(rhs(template.ast`x = function () { const x = 10; return x; }`)))
```

JoostK commented on packages/compiler-cli/linker/test/ast/babel/babel_ast_host_spec.ts:
```diff
@@ -0,0 +1,282 @@
+              host.parseReturnValue(rhs(template.ast`x = function () { const x = 10; return x;}`)))
+          .toThrowError(
+              'Unsupported syntax, expected a function body with a single return statement.');
+      expect(() => host.parseReturnValue(rhs(template.ast`x = () => { const x = 10; return x;}`)))
```
```suggestion
      expect(() => host.parseReturnValue(rhs(template.ast`x = () => { const x = 10; return x; }`)))
```

JoostK commented on packages/compiler-cli/linker/test/ast/typescript/typescript_ast_host_spec.ts:
```diff
@@ -0,0 +1,274 @@
+      expect(host.isNumericLiteral(expr(`[]`))).toBe(false);
+      expect(host.isNumericLiteral(expr(`null`))).toBe(false);
+      expect(host.isNumericLiteral(expr(`'a' + 'b'`))).toBe(false);
+      expect(host.isNumericLiteral(expr(`\`moo\``))).toBe(false);
```
No tagged template here so this could be:

```suggestion
      expect(host.isNumericLiteral(expr('`moo`'))).toBe(false);
```

In fact, many of the template strings in this file can be regular string literals.

petebacondarwin replied:
Yeah they are leftover from when all of these were tagged template literals.

petebacondarwin replied:
I have removed all the backticks

JoostK commented on packages/compiler-cli/linker/src/ast/babel/babel_ast_factory.ts:
```diff
@@ -0,0 +1,166 @@
+      leftOperand: t.Expression, operator: BinaryOperator,
+      rightOperand: t.Expression): t.Expression {
+    switch (operator) {
+      case '&&':
+      case '||':
+        return t.logicalExpression(operator, leftOperand, rightOperand);
```
I don't think I've seen tests for these. Although tests probably won't add much value beyond the type checker yelling if this was incorrect.

petebacondarwin replied:
I'll add one for good measure

JoostK commented on packages/compiler-cli/linker/test/linker_import_generator_spec.ts:
```diff
@@ -0,0 +1,39 @@
+ */
+import {LinkerImportGenerator} from '../src/linker_import_generator';
+
+const ngImport = {};
```
If this has some properties it's harder to trick the `toEqual` into accepting another empty object.

```suggestion
const ngImport = {ngImport: true};
```

petebacondarwin commented on packages/compiler-cli/linker/src/ast/babel/babel_ast_factory.ts:
```diff
@@ -0,0 +1,166 @@
+
+  createTaggedTemplate(tag: t.Expression, template: TemplateLiteral<t.Expression>): t.Expression {
+    const elements = template.elements.map(
+        (element, i) => t.templateElement(element, i === template.elements.length - 1));
```
Good point!
This never got updated after all the upheaval in the `TypescriptAstFactory` around the `range` stuff.

petebacondarwin commented on packages/compiler-cli/linker/src/ast/babel/babel_ast_factory.ts:
```diff
@@ -0,0 +1,166 @@
+    if (sourceMapRange === null) {
+      return node;
+    }
+    // TODO: Babel does not appear to support external source map sources. We should verify that
+    //  the range that is being set at least corresponds with the file that is transformed,
+    //  otherwise the line/column information should not used to set the node's location.
```
I have been procrastinating about this and planned to do some runtime digging around.
It is possible to have the filename assigned to ranges in Babel...

petebacondarwin replied:
Removed this and left a comment

petebacondarwin commented on packages/compiler-cli/linker/src/ast/babel/babel_ast_factory.ts:
```diff
@@ -0,0 +1,166 @@
+    //  otherwise the line/column information should not used to set the node's location.
+    node.loc = {
+      start: {
+        line: sourceMapRange.start.line,
```
Good spot!

petebacondarwin committed:
fixup! refactor(compiler-cli): implement `BabelAstFactory` and `AstHost`s

petebacondarwin committed:
fixup! refactor(compiler-cli): implement `BabelAstFactory` and `AstHost`s

petebacondarwin committed:
fixup! refactor(compiler-cli): implement `BabelAstFactory` and `AstHost`s

devversion reviewed (approved):
LGTM for the dev-infra affected files.

petebacondarwin committed:
test(compiler-cli): make typescript_ast_factory_spec tests resilient to line-endings

The tests were assuming that newlines were `\n` characters but this is not
the case on Windows. This was fixed in #38925, but a better solution is to
configure the TS printer to always use `\n` characters for newlines.

gkalpak commented on packages/compiler-cli/linker/src/ast/babel/babel_ast_factory.ts:
```diff
@@ -0,0 +1,165 @@
+
+  createExpressionStatement = t.expressionStatement;
+
+  createFunctionDeclaration(functionName: string|null, parameters: string[], body: t.Statement):
```
Can `functionName` really be `null`? Can a function declaration be anonymous?

petebacondarwin replied:
You are correct. It cannot be `null`. I think this was cut-n-pasted from `createFunctionExpression()` where it can be `null`.

JoostK replied:
It can be:

```ts
export default function() {}
```

But we don't have to support it, in fact we don't support generating the `export default`, so it's fine to require a name.

petebacondarwin replied:
Ahah! I forgot about that form. Let's not support it for now. It is easy enough to add in if required and it keeps our code simpler.

gkalpak commented on packages/compiler-cli/linker/test/ast/babel/babel_ast_factory_spec.ts:
```diff
@@ -0,0 +1,382 @@
+      expect(expr.end).toBeUndefined();
+
+      factory.setSourceMapRange(expr, {
+        start: {line: 0, column: 1, offset: 2},
```
How can `offset` be 2 if `line` is 0 and `column` is 1? 😕 :grin:

petebacondarwin replied:
I was just trying to use numbers that would be distinct :-)
Actually in the TypeScript version the positions need to match the actual source content. So I will realign this test with the TS one.

gkalpak reviewed (changes requested):
Some minor comments and questions. Otherwise lgtm 🚀

petebacondarwin commented on packages/compiler-cli/linker/index.ts:
```diff
@@ -0,0 +1,9 @@
+ * found in the LICENSE file at https://angular.io/license
+ */
+
+export {FatalLinkerError, isFatalLinkerError} from './src/fatal_linker_error';
```
I was considering whether we might need to use this in a standalone binary.
But I guess we can add it as needed - so I will remove it now.

JoostK commented on .github/angular-robot.yml:
```diff
@@ -48,6 +48,7 @@ merge:
       - "packages/bazel/src/protractor/**"
       - "packages/bazel/src/schematics/**"
       - "packages/compiler-cli/ngcc/**"
+      - "packages/compiler-cli/linker/**"
```
btw, I think this should be before ngcc per the alphabet.

petebacondarwin commented on packages/compiler-cli/linker/src/ast/babel/babel_ast_factory.ts:
```diff
@@ -0,0 +1,165 @@
+}
+
+function isLExpression(expr: t.Expression): expr is Extract<t.LVal, t.Expression> {
+  // Some LVal types are not expressions, which prevents us from using `t.isLVal()`
+  // directly with `assert()`.
```
Because the `expr` is already typed to `t.Expression`, which constrains the return value because in a type guard like

```ts
function isX(val: Y): val is X;
```

it must be true that `X extends Y`

petebacondarwin replied:
And if `expr` is already a `t.Expression` then the type guard of `t.isVal(expr)` is narrowed.

petebacondarwin commented on packages/compiler-cli/linker/src/ast/babel/babel_ast_factory.ts:
```diff
@@ -0,0 +1,165 @@
+
+  createTaggedTemplate(tag: t.Expression, template: TemplateLiteral<t.Expression>): t.Expression {
+    const elements = template.elements.map(
+        (element, i) => this.setSourceMapRange(
```
I might be wrong... but the expressions will have already been visited themselves by the translator, so they will have had their ranges applied already.

We can't rely upon this happening for elements because in the ES5 version they get split up and converted into separate arrays.

petebacondarwin committed:
refactor(compiler-cli): function declarations must have names

The `AstFactory.createFunctionDeclaration()` was allowing `null` to be
passed as the function `name` value. This is not actually possible, since
function declarations must always have a name.

JoostK commented:
Reminder: we need to add a dependency on Babel in `@angular/compiler-cli`'s `package.json`.

petebacondarwin commented:
> Reminder: we need to add a dependency on Babel in `@angular/compiler-cli`'s `package.json`.

https://github.com/angular/angular/pull/38866/commits/05a233f1240bd3c9a325a558ac77f73456764eea

gkalpak reviewed (approved):
:shipit:

AndrewKushnir commented:
[Presubmit](http://test/OCL:334622060:BASE:334622088:1601485349616:d260e478).

alxhub commented on packages/compiler-cli/linker/src/ast/babel/babel_ast_factory.ts:
```diff
@@ -0,0 +1,165 @@
+import {assert} from '../utils';
+
+export class BabelAstFactory implements AstFactory<t.Statement, t.Expression> {
+  attachComments(statement: t.Statement, leadingComments?: LeadingComment[]): t.Statement {
```
I see this is the `AstFactory` interface, but it's confusing to me that a method known as `attachComments` only optionally takes comments to attach.

petebacondarwin replied:
Yes. I'll change it to `LeadingComment[]|undefined` throughout.

alxhub commented on packages/compiler-cli/linker/src/ast/babel/babel_ast_factory.ts:
```diff
@@ -0,0 +1,165 @@
+      return node;
+    }
+    // Note that the linker only works on a single file at a time, so there is no need to track the
+    // filename. Babel will just use the current filename in the source-map.
```
🎉

alxhub commented on packages/compiler-cli/linker/src/ast/babel/babel_ast_host.ts:
```diff
@@ -0,0 +1,139 @@
+import {AstHost, Range} from '../ast_host';
+import {assert} from '../utils';
+
+export class BabelAstHost implements AstHost<t.Expression> {
```
Doc comment :)

alxhub commented on packages/compiler-cli/linker/src/ast/typescript/typescript_ast_host.ts:
```diff
@@ -0,0 +1,145 @@
+
+
+/**
+ * This host is not actually used at runtime in the current code.
```
Add a first line describing what this class actually is (implementation of `AstHost` for the TS AST), before this disclaimer.

alxhub commented on packages/compiler-cli/linker/test/ast/babel/babel_ast_factory_spec.ts:
```diff
@@ -0,0 +1,382 @@
+
+  describe('createBinaryExpression()', () => {
+    it('should create a binary operation node using the left and right expressions', () => {
+      const left = expression.ast`17`;
```
Wow, this is such a cool use of tagged template functions!

petebacondarwin replied:
Yeah, Babel has some very nice DX

alxhub commented:
[Presubmit](https://test.corp.google.com/ui#id=OCL:334643463:BASE:334643496:1601490857524:9d667573)
